utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/universal/ConfluentUniversalKafkaDeserializerFactory.scala

```diff
@@ -3,30 +3,82 @@ package pl.touk.nussknacker.engine.avro.schemaregistry.confluent.serialization.u
+import pl.touk.nussknacker.engine.avro.schemaregistry.confluent.serialization.ConfluentAvroPayloadDeserializer
+import pl.touk.nussknacker.engine.avro.schemaregistry.confluent.serialization.universal.ConfluentUniversalKafkaSerde.SchemaIdHeaderName
 import pl.touk.nussknacker.engine.avro.serialization.KafkaSchemaBasedKeyValueDeserializationSchemaFactory
 import pl.touk.nussknacker.engine.kafka.KafkaConfig
 
+import java.nio.ByteBuffer
 import scala.reflect.ClassTag
+import scala.util.Try
+
+object ConfluentUniversalKafkaSerde {
+  val SchemaIdHeaderName = "schema_id"
+}
+
+class ConfluentUniversalKafkaDeserializer[T](
+                                schemaRegistryClient: ConfluentSchemaRegistryClient,
+                                readerSchemaDataOpt: Option[RuntimeSchemaData[ParsedSchema]],
+                                isKey: Boolean
+                              ) extends Deserializer[T] {
+  private case class WithShiftingRequiredInfo[A](get: A, shouldShiftMagicByte: Boolean) {
+    def map[B](fn: A => B): WithShiftingRequiredInfo[B] = WithShiftingRequiredInfo(fn(get), shouldShiftMagicByte)
+    def bufferStartPosition: Int = if(shouldShiftMagicByte) 1 + ConfluentUtils.IdSize else 1
+  }
+
+  override def deserialize(topic: String, data: Array[Byte]): T = {
+    throw new IllegalAccessException()
+  }
+
+  private lazy val avroPayloadDeserializer = new ConfluentAvroPayloadDeserializer(false, false, false, DecoderFactory.get())
+
+  override def deserialize(topic: String, headers: Headers, data: Array[Byte]): T = {
+    val writerSchema = getParsedSchema(headers, data)
+    readerSchemaDataOpt.map(_.schema.schemaType()).foreach(readerSchemaType => {
+      if(readerSchemaType != writerSchema.get.schemaType())
+        throw new IllegalArgumentException(s"Expecting schema of type $readerSchemaType. but got payload with ${writerSchema.get.schemaType()} schema type")
+    })
+
+    writerSchema.get match {
+      //todo handle JsonSchema
+      case schema: AvroSchema =>
+        val writerAvroSchema = RuntimeSchemaData(schema.rawSchema(), Some(schema.version().toInt))
+        val readerAvroSchema = readerSchemaDataOpt.asInstanceOf[Option[RuntimeSchemaData[AvroSchema]]]
+        avroPayloadDeserializer.deserialize(readerAvroSchema, writerAvroSchema, ByteBuffer.wrap(data), writerSchema.bufferStartPosition).asInstanceOf[T]
+      case _ => throw new IllegalArgumentException("Not supported schema type")
+    }
+
+  }
+
+  private def getParsedSchema(headers: Headers, data: Array[Byte]): WithShiftingRequiredInfo[ParsedSchema] = {
+    val schemaId = Option(headers.lastHeader(SchemaIdHeaderName)) match {
+      case Some(header) =>
+        val id = Try(new String(header.value()).toInt)
+          .fold(e => throw new IllegalArgumentException(s"Got header $SchemaIdHeaderName, but the value is corrupted.", e), x => x)
```

Comment:
Can u print that value? Also not exactly corrupted. It is rather "invalid" value, you don't know if someone broken it or by design send not supported value e.g. uuid